src/services/authorization.rs

```diff
@@ -1,199 +1,142 @@
 use std::sync::Arc;
 
+use casbin::prelude::*;
+use serde::{Deserialize, Serialize};
+use tokio::sync::RwLock;
+
 use super::user::Repository;
 use crate::errors::ServiceError;
 use crate::models::user::{UserCompact, UserId};
 
+#[derive(Debug, Clone, Serialize, Deserialize, Hash)]
 pub enum ACTION {
     AddCategory,
     DeleteCategory,
     GetSettings,
     GetSettingsSecret,
     AddTag,
     DeleteTag,
+    DeleteTorrent,
+    BanUser,
 }
 
 pub struct Service {
     user_repository: Arc<Box<dyn Repository>>,
+    casbin_enforcer: Arc<CasbinEnforcer>,
 }
 
 impl Service {
     #[must_use]
-    pub fn new(user_repository: Arc<Box<dyn Repository>>) -> Self {
-        Self { user_repository }
+    pub fn new(user_repository: Arc<Box<dyn Repository>>, casbin_enforcer: Arc<CasbinEnforcer>) -> Self {
+        Self {
+            user_repository,
+            casbin_enforcer,
+        }
     }
 
+    ///Allows or denies an user to perform an action based on the user's privileges
+    ///
     /// # Errors
     ///
     /// Will return an error if:
-    ///
-    /// - There is not any user with the provided `UserId` (when the user id is some).
+    /// - There is no user id found in the request
+    /// - The user id is not found in the database
     /// - The user is not authorized to perform the action.
-    pub async fn authorize(&self, action: ACTION, maybe_user_id: Option<UserId>) -> Result<(), ServiceError> {
-        match action {
-            ACTION::AddCategory
-            | ACTION::DeleteCategory
-            | ACTION::GetSettings
-            | ACTION::GetSettingsSecret
-            | ACTION::AddTag
-            | ACTION::DeleteTag => match maybe_user_id {
-                Some(user_id) => {
-                    let user = self.get_user(user_id).await?;
-
-                    if !user.administrator {
-                        return Err(ServiceError::Unauthorized);
-                    }
 
+    pub async fn authorize(&self, action: ACTION, maybe_user_id: Option<UserId>) -> std::result::Result<(), ServiceError> {
+        match maybe_user_id {
+            Some(user_id) => {
+                // Checks if the user found in the request exists in the database
+                let user_guard = self.get_user(user_id).await?;
+
+                //Converts the bool administrator value to a string so the enforcer can handle the request and match against the policy file
+                let role = if user_guard.administrator { "admin" } else { "guest" };
+
+                let enforcer = self.casbin_enforcer.enforcer.read().await;
+
+                let authorize = enforcer.enforce((role, action)).map_err(|_| ServiceError::Unauthorized)?;
+
+                if authorize {
                     Ok(())
+                } else {
+                    Err(ServiceError::Unauthorized)
                 }
-                None => Err(ServiceError::Unauthorized),
-            },
+            }
+            None => Err(ServiceError::Unauthorized),
```

Comment:
Hi @mario-nt, remember that some actions, like listing torrents and viewing torrent details, are not included yet, and they don't require a logged-in user. So this case should also be handled by the enforcer. If we don't know the user role we should pass the "unknown" role to casbin. This won't work in the future when we include the actions you can perform as a guest user.

Reply:
True, I was planning on addressing this issue with the guest users/public actions handlers on the PR that I will open specifically for those actions.

I was thinking on either refactor the current authorize method or create a new one called authorized_public handlers or something similar.

I will work on those changes once I finish this PR and start working on the one for the public actions handlers.